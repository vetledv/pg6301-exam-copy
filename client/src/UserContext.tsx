import { createContext, useState } from 'react'

export type userContextType = 'anon' | 'reader' | 'editor'

export interface UserContextState {
    userType: userContextType
    setUserType: (user: userContextType) => void
}

// This context could've been done with the useUser custom hook
// to check for user object microsoft, google or {},
// but I wanted to showcase using context as well
export const UserContext = createContext<UserContextState | null>(null)

export const UserContextProvider = ({ children }: any) => {
    const [userType, setUserType] = useState<userContextType>('anon')
    return (
        <UserContext.Provider value={{ userType, setUserType }}>
            {children}
        </UserContext.Provider>
    )
}
