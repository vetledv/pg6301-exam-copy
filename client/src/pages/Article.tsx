import { useNavigate, useParams } from 'react-router-dom'
import { ArticleLayoutSingle } from '../components/ArticleLayouts'
import { useArticle } from '../hooks/reactQueryHooks'

export const Article = () => {
    const navigate = useNavigate()
    const { id } = useParams()
    const article = useArticle(id as string)
    if (article.isLoading) {
        return <div>Loading...</div>
    }
    if (article.isError) {
        return <div>Error: {article.error.message}</div>
    }

    return (
        <div className='w-full grid grid-cols-1 lg:grid-cols-4 gap-2'>
            <button
                className='px-6 py-2 rounded-lg bg-cyan-600 w-fit text-contrast dark:text-primary col-span-1 lg:col-span-3'
                onClick={() => navigate(-1)}>
                Back
            </button>
            <ArticleLayoutSingle article={article.data!} />
        </div>
    )
}
