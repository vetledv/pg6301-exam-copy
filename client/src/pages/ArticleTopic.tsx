import { useParams } from 'react-router-dom'
import { ArticleLayoutHome } from '../components/ArticleLayouts'
import { useAllArticlesByTopic } from '../hooks/reactQueryHooks'

export const ArticleTopic = () => {
    const { topic } = useParams() as { topic: string }
    const uTopic = topic?.charAt(0).toUpperCase() + topic?.slice(1)
    const articles = useAllArticlesByTopic(topic)
    if (articles.isLoading) {
        return <div>Loading...</div>
    }
    if (articles.isError) {
        return <div>Error: {articles.error.message}</div>
    }
    if (articles.data?.length === 0) {
        return <div>No articles found</div>
    }
    return (
        <div className='flex flex-col gap-2 w-full lg:full xl:w-3/4'>
            <div className='text-xl'>{uTopic}</div>
            {articles.data &&
                articles.isSuccess &&
                articles.data
                    .slice()
                    .reverse()
                    .map((article, i) => (
                        <ArticleLayoutHome key={i} article={article} />
                    ))}
        </div>
    )
}
