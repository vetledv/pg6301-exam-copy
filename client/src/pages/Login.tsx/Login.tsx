import { CgMicrosoft } from 'react-icons/cg'
import { FcGoogle } from 'react-icons/fc'
import { useNavigate } from 'react-router-dom'
import { LoginButton } from '../../components/LoginButton'
import { useUser } from '../../hooks/reactQueryHooks'

export const Login = () => {
    const user = useUser()
    const navigate = useNavigate()

    if (user.isLoading || !user.data) {
        return <div>Loading...</div>
    }
    if (user.isError) {
        return <div>Error: {user.error.message}</div>
    }

    return (
        <div className='flex justify-center p-6 w-full sm:w-2/3 sm:max-w-md m-auto rounded-md sm:shadow-md'>
            <div className='flex flex-col w-full gap-4'>
                <h1 className='text-lg'>Log in with:</h1>
                <div className='flex flex-col gap-2'>
                    <LoginButton
                        classname=' bg-gray-100 text-black hover:bg-gray-200 transition-colors shadow'
                        config={user.data.config}
                        label={'Google'}
                        provider={'google'}
                        icon={<FcGoogle className='h-6 w-6' />}
                    />
                    <LoginButton
                        classname='bg-[#0078D4] hover:bg-[#0062cc] transition-colors shadow'
                        config={user.data.config}
                        label={'Microsoft'}
                        provider={'microsoft'}
                        icon={<CgMicrosoft className='h-6 w-6' />}
                    />
                </div>
                <div
                    onClick={() => {
                        navigate(-1 || '/')
                    }}
                    className=' w-fit self-end cursor-pointer'>
                    Go back
                </div>
            </div>
        </div>
    )
}
