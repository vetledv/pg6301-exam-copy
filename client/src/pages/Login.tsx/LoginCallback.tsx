import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { Loading } from '../../components/Loading'
import { useMutateUser, useUser } from '../../hooks/reactQueryHooks'
import { loginUser } from '../../lib/fetch'

export const LoginCallback = () => {
    const user = useUser()
    const { provider } = useParams()
    const [error, setError] = useState<string | null>(null)
    const userMutate = useMutateUser(loginUser, provider)

    useEffect(() => {
        const handleLogin = async () => {
            const { access_token, error, error_description, state, code } =
                Object.fromEntries(
                    new URLSearchParams(window.location.hash.substring(1))
                )
            const expected_state =
                window.sessionStorage.getItem('expected_state')

            if (!state || expected_state !== state) {
                setError('Invalid state')
            }
            if (error || error_description) {
                setError(`Error: ${error} (${error_description})`)
            }
            if (
                code &&
                user.data?.config &&
                provider === ('microsoft' || 'google')
            ) {
                const { client_id, token_endpoint } = user.data.config[provider]
                const code_verifier = window.sessionStorage.getItem(
                    'code_verifier'
                ) as string
                const parameters = {
                    client_id,
                    grant_type: 'authorization_code',
                    code,
                    code_verifier,
                    redirect_uri: `${window.location.origin}/login/${provider}/callback`,
                }
                const res = await fetch(token_endpoint, {
                    method: 'POST',
                    body: new URLSearchParams(parameters),
                })
                if (!res.ok) {
                    setError(`Fetch token failed ${res.status}: ${res.text()}`)
                    return
                }
                const { access_token } = await res.json()
                userMutate.mutate({ access_token })
                return
            }

            if (!access_token) {
                setError('Missing access_token')
                return
            }
            userMutate.mutate({ access_token })
        }
        handleLogin()
    }, [])

    if (error) {
        return <div>{error.toString()}</div>
    } else {
        return <Loading />
    }
}
