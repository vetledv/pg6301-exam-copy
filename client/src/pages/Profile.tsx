import { useNavigate } from 'react-router-dom'
import { Loading } from '../components/Loading'
import { useMutateUser, useUser } from '../hooks/reactQueryHooks'
import { logoutUser } from '../lib/fetch'

export const Profile = () => {
    const navigate = useNavigate()
    const user = useUser()
    const logout = useMutateUser(logoutUser)

    const handleLogout = async () => {
        logout.mutate(null)
    }

    if (user.isLoading || !user.data) {
        return <Loading />
    }

    if (user.isError) {
        return (
            <div className='text-2xl font-bold py-2'>
                Error: {user.error.message}
            </div>
        )
    }
    const gUser = user.data?.user.google
    const mUser = user.data?.user.microsoft
    if (!gUser && !mUser) {
        return (
            <>
                <div className=''>Not logged in, can't access profile</div>
                <div
                    className='underline cursor-pointer py-2'
                    onClick={() => navigate('/login')}>
                    Log in here
                </div>
            </>
        )
    }

    return (
        <div className='w-full grid grid-cols-1 lg:grid-cols-4'>
            <div className='rounded-lg p-4 bg-secondary grid col-span-1 lg:col-span-3 gap-4 '>
                <div className='flex items-center'>
                    {gUser ? (
                        <img
                            src={gUser ? gUser.picture : mUser?.picture}
                            alt={gUser ? gUser.picture : mUser?.picture}
                            className='rounded-full'></img>
                    ) : (
                        <CircleImage
                            letter={mUser!.name.charAt(0)}></CircleImage>
                    )}
                    <div className='px-4'>
                        <div>{gUser ? gUser.name : mUser?.name}</div>
                        <div>{gUser ? gUser.email : mUser?.email}</div>
                    </div>
                </div>
                <div className='border-b border-secondary' />
                <button
                    className='rounded-lg w-fit px-6 py-2 bg-red-400 hover:bg-red-600 transition-colors '
                    onClick={handleLogout}>
                    Log out
                </button>
            </div>
        </div>
    )
}
const CircleImage = ({ letter }: { letter: string }) => {
    return (
        <div className='w-24 h-24 bg-blue-500 rounded-full flex justify-center items-center text-contrast dark:text-primary'>
            <div className='text-2xl'>{letter}</div>
        </div>
    )
}
