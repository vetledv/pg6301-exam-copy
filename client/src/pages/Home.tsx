import { useContext, useEffect, useRef } from 'react'
import { useInfiniteQuery } from 'react-query'
import { useNavigate } from 'react-router-dom'
import { ArticleLayoutHome } from '../components/ArticleLayouts'
import { NewArticleButton } from '../components/NewArticleButton'
import { InfArticles } from '../hooks/reactQueryHooks'
import SkeletonArticle from '../SkeletonArticle'
import { UserContext, UserContextState } from '../UserContext'

export const Home = () => {
    const navigate = useNavigate()
    const socket = useRef<WebSocket | null>(null)
    const fetchPage = async ({ pageParam = 1 }) => {
        const res = await fetch(`/api/articles?page=${pageParam}`)
        const data = await res.json()
        return data
    }

    const query = useInfiniteQuery<InfArticles, Error>(
        ['articlesInf'],
        fetchPage,
        {
            getNextPageParam: (lastPage, pages) => lastPage.next,
        }
    )
    const { userType } = useContext(UserContext) as UserContextState

    useEffect(() => {
        socket.current = new WebSocket(
            window.location.origin.replace(/^http/, 'ws')
        )
        socket.current.onopen = () => {
            console.log('socket opened')
        }
        socket.current.onmessage = () => {
            query.refetch()
        }
        socket.current.onclose = () => {
            console.log('socket closed')
        }
        const currSocket = socket.current
        return () => {
            currSocket.close()
        }
    }, [])

    if (query.isLoading || !query.data) {
        return (
            <div className='flex flex-col gap-2 w-full lg:full xl:w-3/4'>
                <SkeletonArticle />
            </div>
        )
    }

    if (query.isError) {
        return <div className=' py-2'>Error: {query.error.message}</div>
    }

    return (
        <div
            ref={null}
            className='gap-2 w-full grid grid-cols-1 lg:grid-cols-4'>
            {userType === 'editor' && (
                <div className='  col-span-1 lg:col-span-3 '>
                    <NewArticleButton socket={socket} />
                </div>
            )}
            {userType === 'anon' && (
                <div className='col-span-1 lg:col-span-3'>
                    <div
                        data-testid='login-button-home'
                        className=' w-fit px-4 py-2 rounded-lg bg-secondary cursor-pointer'
                        onClick={() => navigate('/login')}>
                        Log in to see more
                    </div>
                </div>
            )}
            {query.data.pages.map((page) =>
                page.articles.map((article, i) => (
                    <div key={i} className='col-span-1 lg:col-span-3'>
                        <ArticleLayoutHome article={article} />
                    </div>
                ))
            )}
            <div className='col-span-1 lg:col-span-3'>
                {query.hasNextPage ? (
                    <button
                        className='w-full px-4 py-2 rounded-lg bg-secondary cursor-pointer'
                        onClick={() => {
                            if (query.hasNextPage) {
                                query.fetchNextPage()
                            }
                        }}>
                        Load more
                    </button>
                ) : null}
            </div>
        </div>
    )
}
