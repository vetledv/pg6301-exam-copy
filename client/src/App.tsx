import { useContext, useEffect } from 'react'
import { QueryClient, QueryClientProvider } from 'react-query'
import { ReactQueryDevtools } from 'react-query/devtools/development'
import { BrowserRouter, Route, Routes } from 'react-router-dom'
import { Layout } from './components/Layout'
import { useUser } from './hooks/reactQueryHooks'
import { Article } from './pages/Article'
import { ArticleTopic } from './pages/ArticleTopic'
import { Home } from './pages/Home'
import { Login } from './pages/Login.tsx/Login'
import { LoginCallback } from './pages/Login.tsx/LoginCallback'
import { Profile } from './pages/Profile'
import SkeletonArticle from './SkeletonArticle'
import {
    UserContext,
    UserContextProvider,
    UserContextState,
} from './UserContext'

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            refetchOnWindowFocus: false,
            refetchOnMount: false,
            keepPreviousData: true,
        },
    },
})
export const App = () => {
    return (
        <QueryClientProvider client={queryClient}>
            <UserContextProvider>
                <AppRouter />
                <ReactQueryDevtools initialIsOpen={false} />
            </UserContextProvider>
        </QueryClientProvider>
    )
}

const AppRouter = () => {
    const query = useUser()
    const { setUserType } = useContext(UserContext) as UserContextState

    useEffect(() => {
        if (!query.data) {
            setUserType('anon')
        }
        if (query.data?.user.microsoft) {
            setUserType('editor')
        }
        if (query.data?.user.google) {
            setUserType('reader')
        }
    }, [query])

    if (query.isLoading || query.isFetching || query.isRefetching) {
        return (
            <BrowserRouter>
                <Layout>
                    <div className='flex flex-col gap-2 w-full lg:full xl:w-3/4'>
                        <SkeletonArticle />
                        <SkeletonArticle />
                        <SkeletonArticle />
                        <SkeletonArticle />
                        <SkeletonArticle />
                    </div>
                </Layout>
            </BrowserRouter>
        )
    }

    return (
        <BrowserRouter>
            <Layout>
                <Routes>
                    <Route path={'/'} element={<Home />}></Route>
                    <Route path={'/login/'} element={<Login />} />
                    <Route
                        path={'/login/:provider/callback'}
                        element={<LoginCallback />}
                    />
                    <Route path={'/login/*'} element={<Login />} />
                    <Route path={'/article/:id'} element={<Article />} />
                    <Route
                        path={'/articles/topic/:topic'}
                        element={<ArticleTopic />}
                    />
                    <Route path={'/profile'} element={<Profile />} />
                    <Route path='*' element={<div>404 not found</div>} />
                </Routes>
            </Layout>
        </BrowserRouter>
    )
}
