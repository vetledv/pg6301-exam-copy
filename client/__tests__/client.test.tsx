require('../src/mocks/matchMedia.mock')
import '@testing-library/jest-dom'
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'
import { ReactNode } from 'react'
import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { TextDecoder, TextEncoder } from 'util'
import * as index from '../src/index'
import { Article } from '../src/pages/Article'
import { ArticleTopic } from '../src/pages/ArticleTopic'
import { Home } from '../src/pages/Home'
import { Login } from '../src/pages/Login.tsx/Login'
import { LoginCallback } from '../src/pages/Login.tsx/LoginCallback'
import { Profile } from '../src/pages/Profile'
import { UserContext, userContextType } from '../src/UserContext'
require('whatwg-fetch')
global.TextEncoder = TextEncoder
//@ts-ignore
global.TextDecoder = TextDecoder

const createTestQueryClient = () =>
    new QueryClient({
        defaultOptions: {
            queries: {
                retry: false,
            },
        },
    })

//wrap custom hook in wrapper
export const createWrapper = () => {
    const testQueryClient = new QueryClient({
        defaultOptions: {
            queries: {
                retry: false,
            },
        },
    })
    return ({ children }: { children: React.ReactElement }) => (
        <QueryClientProvider client={testQueryClient}>
            {children}
        </QueryClientProvider>
    )
}
//internal function used from react-query
//https://github.com/tannerlinsley/react-query/blob/ead2e5dd5237f3d004b66316b5f36af718286d2d/src/react/tests/utils.tsx#L6-L17
export const renderWithClient = (ui: React.ReactElement) => {
    const testQueryClient = createTestQueryClient()
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>
                    {rerenderUi}
                </QueryClientProvider>
            ),
    }
}

export const RenderWithUserContext = ({
    children,
    userType,
    initialEntries,
}: {
    children: ReactNode
    userType?: userContextType
    initialEntries?: string[]
}) => {
    userType = userType || 'anon'
    const setUserType = jest.fn()
    return (
        <UserContext.Provider value={{ userType, setUserType }}>
            <MemoryRouter initialEntries={initialEntries}>
                {children}
            </MemoryRouter>
        </UserContext.Provider>
    )
}

describe('Client tests', () => {
    //creds: https://stackoverflow.com/questions/71779034/jest-test-create-react-app-index-js-with-new-version-18/71798229#71798229
    test('Render without crashing', () => {
        expect(
            JSON.stringify(
                Object.assign({}, index, { _reactInternalInstance: 'censored' })
            )
        ).toMatchSnapshot()
    })

    test('Home loading', async () => {
        renderWithClient(
            <RenderWithUserContext>
                <Home />
            </RenderWithUserContext>
        )
        const loading = await screen.findByTestId('loading')
        expect(loading).toBeInTheDocument()
        await waitFor(() => {
            expect(screen.getByText(/Log in to see more/i)).toBeInTheDocument()
        })
    })

    test('Article page', async () => {
        renderWithClient(
            <RenderWithUserContext initialEntries={['/69']} userType='editor'>
                <Article />
            </RenderWithUserContext>
        )
        const loading = await screen.findByText(/Loading.../i)
        expect(loading).toBeInTheDocument()
        await waitFor(() => {
            const article = screen.getByText(/Vetle Brandth/i)
            expect(article).toBeInTheDocument()
        })
        await waitFor(() => {
            const editBtn = screen.getByText(/Edit/i)
            expect(editBtn).toBeInTheDocument()
            fireEvent.click(editBtn)
            expect(screen.getByText(/Cancel/i)).toBeInTheDocument()
        })
        await waitFor(() => {
            const headlineInput = screen.getByTestId('edit-headline')
            fireEvent.change(headlineInput, {
                target: { value: 'New headline' },
            })
            expect(headlineInput).toHaveValue('New headline')
            const bodyInput = screen.getByTestId('edit-body')
            fireEvent.change(bodyInput, {
                target: { value: 'New body' },
            })
            expect(bodyInput).toHaveValue('New body')
        })
    })

    test('NewArticleButton', async () => {
        renderWithClient(
            <RenderWithUserContext userType='editor'>
                <Home />
            </RenderWithUserContext>
        )
        await waitFor(() => {
            const newArticleBtn = screen.getByText(/New article/i)
            fireEvent.click(newArticleBtn)
            expect(screen.getByText(/Cancel/i)).toBeInTheDocument()
            const submit = screen.getByTestId('new-article-submit')
            fireEvent.click(submit)
            const errorSpans = screen.getAllByText(/This field is required/i)
            expect(errorSpans[0]).toBeInTheDocument()
        })
        await waitFor(() => {
            const newArticleBtn = screen.getByText(/New article/i)
            fireEvent.click(newArticleBtn)
            expect(screen.getByText(/Cancel/i)).toBeInTheDocument()
            fireEvent.change(screen.getByTestId('new-article-headline'), {
                target: { value: 'Hello World' },
            })
            fireEvent.change(screen.getByTestId('new-article-body'), {
                target: { value: 'Hello World' },
            })
            fireEvent.change(screen.getByTestId('new-article-date'), {
                target: { value: '2020-01-01' },
            })
            fireEvent.change(screen.getByTestId('new-article-topic'), {
                target: { value: 'News' },
            })
            const submit = screen.getByTestId('new-article-submit')
            fireEvent.click(submit)
            expect(screen.getByText(/Cancel/i)).toBeInTheDocument()
        })
    })

    test('Login and click login with google', async () => {
        renderWithClient(
            <RenderWithUserContext userType='anon'>
                <Login />
            </RenderWithUserContext>
        )
        //mock  window.location.href to make sure it's not redirected
        window.location.href = 'http://localhost:3000/'
        await waitFor(() => {
            fireEvent.click(screen.getByText(/Google/i))
            expect(sessionStorage.getItem('expected_state')).toBeTruthy()
        })
    })
    test('LoginCallback missing access token', async () => {
        renderWithClient(
            <RenderWithUserContext userType='anon'>
                <LoginCallback />
            </RenderWithUserContext>
        )
        await waitFor(() => {
            expect(
                screen.getByText(/Missing access_token/i)
            ).toBeInTheDocument()
        })
    })

    test('ArticleTopic page', async () => {
        // set url to /topics/News

        renderWithClient(
            <RenderWithUserContext
                initialEntries={['/topic/news']}
                userType='reader'>
                <ArticleTopic />
            </RenderWithUserContext>
        )
        //wait for loading
        await waitFor(() => {
            expect(screen.getByText(/News/i)).toBeInTheDocument()
        })
    })

    test('Profile page', async () => {
        renderWithClient(
            <RenderWithUserContext userType='editor'>
                <Profile />
            </RenderWithUserContext>
        )
        await waitFor(() => {
            expect(screen.getByText(/Vetle Brandth/i)).toBeInTheDocument()
        })
    })
})
