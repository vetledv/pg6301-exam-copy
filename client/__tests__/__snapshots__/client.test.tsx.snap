// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client tests Render without crashing 1`] = `"{\\"_reactInternalInstance\\":\\"censored\\"}"`;
